torchtitan/checkpoint.py

```diff
@@ -103,24 +103,41 @@ def __init__(
         self,
         model: Union[nn.Module, List[nn.Module]],
         optim: Union[torch.optim.Optimizer, List[torch.optim.Optimizer]],
```

Comment:
What would be the typing for `optim` after supporting `optim_in_bwd`?

Reply:
That would help me to understand the code.

Reply:
Thank you for the comments. `optim` after `optim_in_bwd` would be a `List[Optimizer]`
For the general `optim`, it's based from `torch.optim.AdamW(model.parameters())`
While after `optim_in_bwd`, each `param` in `model.parameters()` would generate an sub optimizer `torch.optim.AdamW(param)`, there's a list with length of `len(model.parameters())` of sub optimizers
The sub optimzier here could call the hook separately to minimize the peak memory in gradient calculation.(to free them earlier)

Normally, the `optim` after `optim_in_bwd` would be a structure like `Dict{param: sub_optim(param)}`, here to simplify the code, just leave as `List[sub_optim]`